### PR TITLE
plugins: conda

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
Port the `conda` plugin from [`snapcraft`](https://github.com/snapcore/snapcraft/blob/main/snapcraft/plugins/v2/conda.py) to `craft-parts`.

(CRAFT-1030)

### CODE REVIEW

File: craft_parts/plugins/conda_plugin.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(CondaPluginProperties, self._options)
+
+        deploy_command = [
+            "CONDA_TARGET_PREFIX_OVERRIDE=/snap/${CRAFT_PROJECT_NAME}/current",
```
Comment (mr-cal) [inline craft_parts/plugins/conda_plugin.py:106]: I took this line from [snapcraft](https://github.com/snapcore/snapcraft/blob/0224b48ea0e1cc7378e1ad392b41466f005ae3e6/snapcraft/plugins/v2/conda.py#L138).  It's limits the flexibility of the plugin, but I've been unable to find any better solution.

**Comment by mr-cal** [comment]
Tested and confirmed conda projects run with snapcraft 7.x, although I will need to submit the following changes for snapcraft:
1. Pass `project_name` to `craft_parts.lifecycle_manager`.
2. Convert underscores to hyphens when writing `meta/snap.yaml`

File: craft_parts/plugins/conda_plugin.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+logger = logging.getLogger(__name__)
+
+
+class CondaPluginProperties(PluginProperties, PluginModel):
+    """The part properties used by the conda plugin."""
+
+    # part properties required by the plugin
+    conda_packages: Optional[List[str]] = None
+    conda_python_version: Optional[str] = None
+    conda_install_prefix: str = "/snap/${CRAFT_PROJECT_NAME}/current"
```
Comment (sergiusens) [inline craft_parts/plugins/conda_plugin.py:35]: is the project name not available through options?

File: craft_parts/plugins/conda_plugin.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(CondaPluginProperties, self._options)
+
+        deploy_command = [
+            f"CONDA_TARGET_PREFIX_OVERRIDE={options.conda_install_prefix}",
+            "conda",
+            "create",
+            "--prefix",
+            "$CRAFT_PART_INSTALL",
```
Comment (sergiusens) [inline craft_parts/plugins/conda_plugin.py:114]: is the CRAFT_PART_INSTALL not available through options?

File: craft_parts/plugins/conda_plugin.py (hunk 2 above)
Comment (sergiusens) [reply craft_parts/plugins/conda_plugin.py:35]: this is probably a snapcraft plugin if this is the default

Comment (cmatsuoka) [reply craft_parts/plugins/conda_plugin.py:35]: As discussed with @sergiusens, craft-parts plugins can't assume the resulting artifact will be a snap. If this is essential for conda to work, this should be a snapcraft plugin instead.

Comment (mr-cal) [reply craft_parts/plugins/conda_plugin.py:35]: How about this logic in `get_build_commands`?
```
        if options.conda_install_prefix:
            install_prefix = options.conda_install_prefix
        elif self._part_info.application_name == "snapcraft":
            install_prefix = f"/snap/{self._part_info.project_name}/current"
        else:
            install_prefix = str(self._part_info.part_install_dir)

        deploy_command = [
            f"CONDA_TARGET_PREFIX_OVERRIDE={install_prefix}",
```

**Comment by sergiusens** [comment]
I would just remove the install prefix default and think of a way to add defaults to the model from the application side

**Comment by sergiusens** [comment]
Closing in favor of https://github.com/snapcore/snapcraft/pull/3703

**Pull request closed without merging**